src/lib/operators/join_index.cpp

```diff
@@ -223,71 +220,132 @@ std::shared_ptr<const Table> JoinIndex::_on_execute() {
+                                      secondary_predicate_evaluator,
+                                      !is_semi_or_anti_join};
+    JoinNestedLoop::_join_two_untyped_segments(*probe_segment, *index_segment, probe_chunk_id, index_chunk_id, params);
   }
+  const auto& index_pos_list_size_post_fallback = _index_pos_list->size();
+  const auto& count_index_positions = index_pos_list_size_post_fallback - index_pos_list_size_pre_fallback;
+  std::fill_n(std::back_inserter(_index_pos_dereferenced), count_index_positions, false);
+  performance_data.chunks_scanned_without_index++;
 }
 
 // join loop that joins two segments of two columns using an iterator for the probe side,
 // and an index for the index side
 template <typename ProbeIterator>
-void JoinIndex::_join_two_segments_using_index(ProbeIterator probe_iter, ProbeIterator probe_end,
-                                               const ChunkID probe_chunk_id, const ChunkID index_chunk_id,
-                                               const std::shared_ptr<BaseIndex>& index) {
+void JoinIndex::_data_join_two_segments_using_index(ProbeIterator probe_iter, ProbeIterator probe_end,
+                                                    const ChunkID probe_chunk_id, const ChunkID index_chunk_id,
+                                                    const std::shared_ptr<BaseIndex>& index) {
+  for (; probe_iter != probe_end; ++probe_iter) {
+    const auto probe_side_position = *probe_iter;
+    const auto index_ranges = _index_ranges_for_value(probe_side_position, index);
+    for (const auto& index_range : index_ranges) {
+      _append_matches(std::get<0>(index_range), std::get<1>(index_range), probe_side_position.chunk_offset(),
+                      probe_chunk_id, index_chunk_id);
+    }
+  }
+}
+
+template <typename ProbeIterator>
+void JoinIndex::_reference_join_two_segments_using_index(
+    ProbeIterator probe_iter, ProbeIterator probe_end, const ChunkID probe_chunk_id, const ChunkID index_chunk_id,
+    const std::shared_ptr<BaseIndex>& index, const std::shared_ptr<const PosList>& reference_segment_pos_list) {
   for (; probe_iter != probe_end; ++probe_iter) {
-    const auto probe_side_value = *probe_iter;
-    if (probe_side_value.is_null()) continue;
+    PosList index_scan_pos_list;
+    const auto probe_side_position = *probe_iter;
+    const auto index_ranges = _index_ranges_for_value(probe_side_position, index);
+    for (const auto& index_range : index_ranges) {
+      std::transform(std::get<0>(index_range), std::get<1>(index_range), std::back_inserter(index_scan_pos_list),
+                     [index_chunk_id](ChunkOffset index_chunk_offset) {
+                       return RowID{index_chunk_id, index_chunk_offset};
+                     });
+    }
+
+    PosList sortable_ref_seg_pos_list;
+    sortable_ref_seg_pos_list.insert(sortable_ref_seg_pos_list.end(), reference_segment_pos_list->begin(),
```

Comment:
Can't we just copy it or sort the pos list itself?

Reply:
We cannot sort the pos list itself since it is const. Additionally, the copy constructor of PosList is implicitly-deleted.
To avoid creating and filling an additional vector for performance purposes, we could use std::const_pointer_cast. But this isn't an elegant solution, either.
```c++
const auto& mutable_ref_seg_pos_list = std::const_pointer_cast<PosList>(reference_segment_pos_list);
    std::sort(mutable_ref_seg_pos_list->begin(), mutable_ref_seg_pos_list->end());
```

Reply:
@mrks should we go with the current solution?

Reply:
Unless there is a significant performance reason to avoid copies, we should avoid modifying pos lists as they might be referenced by more than a single ReferenceSegment.

The PosList is non-copyable so that people don't accidentally do stupid stuff with it. I'd be more than fine with adding a `copy()` method.

Reply:
(Implicit copies in C++ are... not so nice anyway)

https://pbs.twimg.com/media/Bhgkwl2IQAAQ5YG.jpg